worker/firewaller/firewaller.go

```diff
@@ -524,68 +529,107 @@ func (fw *Firewaller) flushUnits(unitds []*unitData) error {
 
 // flushMachine opens and closes ports for the passed machine.
 func (fw *Firewaller) flushMachine(machined *machineData) error {
-	// Gather ports to open and close.
-	want := []network.PortRange{}
-	for portRange, unitTag := range machined.definedPorts {
-		unitd, known := machined.unitds[unitTag]
-		if !known {
-			delete(machined.unitds, unitTag)
-			continue
-		}
-		if unitd.serviced.exposed {
-			want = append(want, portRange)
-		}
+	want, err := fw.gatherIngressRules(machined)
+	if err != nil {
+		return errors.Trace(err)
 	}
-	toOpen := diffRanges(want, machined.openedPorts)
-	toClose := diffRanges(machined.openedPorts, want)
-	machined.openedPorts = want
+	toOpen, toClose := diffRanges(machined.ingressRules, want)
+	machined.ingressRules = want
 	if fw.globalMode {
 		return fw.flushGlobalPorts(toOpen, toClose)
 	}
 	return fw.flushInstancePorts(machined, toOpen, toClose)
 }
 
+// gatherIngressRules returns the ingress rules to open and close
+// for the specified machines.
+func (fw *Firewaller) gatherIngressRules(machines ...*machineData) ([]network.IngressRule, error) {
+	var want []network.IngressRule
+	for _, machined := range machines {
+		cidrsForUnit := make(map[names.UnitTag]set.Strings)
+		for unitTag, portRanges := range machined.definedPorts {
+			unitd, known := machined.unitds[unitTag]
+			if !known {
+				delete(machined.unitds, unitTag)
+				continue
+			}
+
+			// Gather the required source CIDRs for the current unit.
+			cidrs, ok := cidrsForUnit[unitTag]
+			if !ok {
+				cidrs = set.NewStrings()
+				cidrsForUnit[unitTag] = cidrs
+			}
+			// If the unit is exposed, allow access from everywhere.
+			if unitd.serviced.exposed {
+				cidrs.Add("0.0.0.0/0")
+			}
+
+			// Add any ingress rules required by remote relations.
```

Comment:
machined.definedPorts *used* to be a map keyed on port range. I changed it to a map keyed on unit tag. So the outer loop:
for unitTag, portRanges := range machined.definedPorts {
}

iterates once per unit. Inside that loop we gather the cidrs required by exposed=true and CMR. Then we construct rules from the set of port ranges for that unit. ie

for unitTag, portRanges := range machined.definedPorts {
  1. gather cidrs for unit tag
  2. construct rules using cidrs and port ranges
}

I think the logic is correct?